Firmware/speed_lookuptable.h

```diff
@@ -39,61 +39,70 @@ FORCE_INLINE uint16_t MUL8x16R8(uint8_t x, uint16_t y) {
     return out;
 }
 
-// intRes = longIn1 * longIn2 >> 24
-// uses:
-// r26 to store 0
-// r27 to store the byte 1 of the 48bit result
-#define MultiU24X24toH16(intRes, longIn1, longIn2) \
-asm volatile ( \
-"clr r26 \n\t" \
-"mul %A1, %B2 \n\t" \
-"mov r27, r1 \n\t" \
-"mul %B1, %C2 \n\t" \
-"movw %A0, r0 \n\t" \
-"mul %C1, %C2 \n\t" \
-"add %B0, r0 \n\t" \
-"mul %C1, %B2 \n\t" \
-"add %A0, r0 \n\t" \
-"adc %B0, r1 \n\t" \
-"mul %A1, %C2 \n\t" \
-"add r27, r0 \n\t" \
-"adc %A0, r1 \n\t" \
-"adc %B0, r26 \n\t" \
-"mul %B1, %B2 \n\t" \
-"add r27, r0 \n\t" \
-"adc %A0, r1 \n\t" \
-"adc %B0, r26 \n\t" \
-"mul %C1, %A2 \n\t" \
-"add r27, r0 \n\t" \
-"adc %A0, r1 \n\t" \
-"adc %B0, r26 \n\t" \
-"mul %B1, %A2 \n\t" \
-"add r27, r1 \n\t" \
-"adc %A0, r26 \n\t" \
-"adc %B0, r26 \n\t" \
-"lsr r27 \n\t" \
-"adc %A0, r26 \n\t" \
-"adc %B0, r26 \n\t" \
-"clr r1 \n\t" \
-: \
-"=&r" (intRes) \
-: \
-"d" (longIn1), \
-"d" (longIn2) \
-: \
-"r26" , "r27" \
-)
+// return ((x * y) >> 24) with rounding when shifting right
+FORCE_INLINE uint16_t MUL24x24R24(__uint24 x, __uint24 y) {
+    uint16_t out;
+    __asm__ (
+    // %0 out
+    // %1 x
+    // %2 y
+    // uint8_t: %An or %n
+    // uint16_t: %Bn %An
+    // __uint24: %Cn %Bn %An
+    // uint32_t: %Dn %Cn %Bn %An
+    //
+    //
+    //    B2 A2 *
+    //       A1
+    //---------
+    // B0 A0 RR
+    "clr r26 \n\t"
+    "mul %A1, %B2 \n\t"
+    "mov r27, r1 \n\t"
+    "mul %B1, %C2 \n\t"
+    "movw %A0, r0 \n\t"
+    "mul %C1, %C2 \n\t"
+    "add %B0, r0 \n\t"
+    "mul %C1, %B2 \n\t"
+    "add %A0, r0 \n\t"
+    "adc %B0, r1 \n\t"
+    "mul %A1, %C2 \n\t"
+    "add r27, r0 \n\t"
+    "adc %A0, r1 \n\t"
+    "adc %B0, r26 \n\t"
+    "mul %B1, %B2 \n\t"
+    "add r27, r0 \n\t"
+    "adc %A0, r1 \n\t"
+    "adc %B0, r26 \n\t"
+    "mul %C1, %A2 \n\t"
+    "add r27, r0 \n\t"
+    "adc %A0, r1 \n\t"
+    "adc %B0, r26 \n\t"
+    "mul %B1, %A2 \n\t"
+    "add r27, r1 \n\t"
+    "adc %A0, r26 \n\t"
+    "adc %B0, r26 \n\t"
+    "lsl r27 \n\t"
```

Comment:
please explain the change of `lsr` to `lsl`

Reply:
@DRracer This is my conversation with Yuri about this

--leptun
Hello :wave:
I was looking at the speed lookup table asm functions and noticed some weird behavior:
```
	MultiU16X8toH16(x, 100, 42); // 0x1068 >>8 => 0x10
	MultiU16X8toH16(x, 101, 42); // 0x1092 >>8 => 0x10
	MultiU16X8toH16(x, 101, 43); // 0x10f7 >>8 => 0x11 ???
	MultiU16X8toH16(x, 95, 41);  // 0x0f37 >>8 => 0x10 ???
	MultiU16X8toH16(x, 95, 42);  // 0x0f96 >>8 => 0x0f ???
```
Why are odd numbers rounded up and even numbers rounded down? That really makes no sense. Is the function trying to round the >>8 but using the lsb instead of the msb by mistake?

--Yuri D'Elia
Huh, good question. No idea, we really need to check the asm. This sounds almost intentional :thinking_face:
Were you trying to use them for some other multiplication?

--leptun
No. I noticed I could optimize it a bit more
And while doing that I also noticed this odd behaviour
Considering this function is used for stepper timer calculation, it made me wonder if this weird rounding is introducing artifacts in accelerations

--Yuri D'Elia
These come from marlin 1.x (even before the MK2 fork)

--leptun
So they might be wrong

Yuri D'Elia
It could be. But at some point when I added the "c" code I did some for loops comparing the outputs and I didn't notice this odd rounding
I remember definitely the comment above the function was wrong. And I only understood what was it attempted to do by cloning marlin and trying to look at the history

--leptun
the c code does absolutely no rounding. So that's what's even weirder
It does a multiply and then right shifts the result. But it also fiddles with this least significant bit for some reason
Btw, here is my proposed optimization:
![image](https://user-images.githubusercontent.com/17808203/153364971-cdb3a020-06cf-4dc0-b289-e90ba75a0eae.png)
Do you see anything that could go wrong by doing so?
The old code was adding r1 to the result and then adding that rounding as the carry of "lsr". r26 was 0, so they were just adding the carry to the result
My optimization simply adds the carry at the same time r1 is added to the result, avoiding those "adc r26" commands
Both my code and the old code exhibit this weird rounding issue
If instead of "lsr r0" I do "lsl r0", then that will get the most significant bit of r0, so it actually rounds the output correctly
lsl and lsr are logical shift btw. They push either the msb or the lsb to the carry and shift the register by 1.
![image](https://user-images.githubusercontent.com/17808203/153365208-82f22f83-cdd3-4a63-a53d-2132060832db.png)
![image](https://user-images.githubusercontent.com/17808203/153365221-fbd1dfe4-4137-4f4e-b21b-cea3edffeab4.png)

Comparison:
```
	asm("nop\n\t");//                             lsr       lsl
	MultiU16X8toH16(x, 100, 42); // 0x1068 >>8 => 0x10      0x10
	MultiU16X8toH16(x, 101, 42); // 0x1092 >>8 => 0x10      0x11
	MultiU16X8toH16(x, 101, 43); // 0x10f7 >>8 => 0x11 ???  0x11
	MultiU16X8toH16(x, 95, 41);  // 0x0f37 >>8 => 0x10 ???  0x0f
	MultiU16X8toH16(x, 95, 42);  // 0x0f96 >>8 => 0x0f ???  0x10
```
We could also technically disable the rounding completely and do just like the C code, but I think doing the rounding correctly might help, unlike the broken rounding which could oscillate between two values with the right conditions

--leptun
Look how it oscillates in this scenario (with lsr like before)
![image](https://user-images.githubusercontent.com/17808203/153365431-0d70b951-41d4-4589-aa1b-2b00e238e942.png)
And this is the same, but with lsl instead of lsr
![image](https://user-images.githubusercontent.com/17808203/153365519-3e0f6418-1ea0-4794-b7a4-ac6dbdaf5b55.png)

MultiU24X24toH16 also oscillates
![image](https://user-images.githubusercontent.com/17808203/153365740-f3f31e5a-4659-46c5-9305-11a181682e40.png)
And, not surprising at all, changing the lsr there to lsl also fixes those oscillations:
![image](https://user-images.githubusercontent.com/17808203/153365785-a3e74c1a-f3f9-49bb-8edb-1388040ce8a4.png)

--leptun
I did two test prints, the surface looks the same :confused:

Reply:
Please keep in mind that I managed to optimize the 16x8 function even more in the meantime, but the stuff mentioned above still holds